pandas/tests/indexing/test_iloc.py

```diff
@@ -1189,6 +1189,18 @@ def test_iloc_getitem_setitem_fancy_exceptions(self, float_frame):
             # GH#32257 we let numpy do validation, get their exception
             float_frame.iloc[:, :, :] = 1
 
+    @pytest.mark.parametrize(
+        "ser, keys",
+        [(Series([10]), (0, 0)), (Series([1, 2, 3], index=list("abc")), (0, 1))],
+    )
+    def test_iloc_setitem_indexer_length(self, ser, keys):
+        # GH#13831
+        with pytest.raises(IndexError, match="too many indices for array"):
```

Comment:
Note the difference in error between `ser.iloc[keys] = 1000` and `ser.iloc[keys]`. `IndexError` is the result of the fact that we let numpy handle this, while pandas raises `IndexingError`. I was wondering if this should be made consistent? Us catching and re-raising from numpy seems ugly.

Reply:
cc @jbrockmendel 

but @DriesSchaumont I think we would want to catch that case & re-raise appropriately

Reply:
IIRC a while back we removed some checks (for perf) and let numpy exceptions surface.

Not knowing anything about the specific use case, in cases where we have a choice between raising IndexError vs IndexingError (i.e. either would be a reasonable choice) I'd rather raise IndexError.